### ISSUE TITLE
flight mode fade in / fade out not applied to Vars

### ISSUE THREAD

**Issue by strgaltdel** [FrSkyRC/ETHOS-Feedback-Community#3142]
we had discussions about implementing fading between FMs in early days
https://github.com/FrSkyRC/ETHOS-Feedback-Community/issues/93

With the introduction of the "Var-Mixer", these parameters were applied to the corresponding channel outputs.

But now, in rev 1.5, the new "GVARs" are not taken into account,
therefore switching between FMs will create "hard" transitions in the mixers/channels using the new Var parameters (in case they are not used in the input section where you can assign special timing parameters)

please implement fading on VARs

**Comment by bsongis-frsky** [comment]
Fading is done on channels, if you need it you should have a free mix which applies the fading
I am not sure I understand your scenario though

**Comment by strgaltdel** [comment]
i' ll try to explain it by a use case

the template in my post simulates camber settings of a typical competition model
(drilled down to minimum settings for the issue)

You set fixed camber positions for all the flaps (&ails) in one mixer by using GVars which have FM dependant values 
(mix type camber, input maximum)

this is for FM neutral/standard:![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/20392683/0d9548e6-3239-4ed6-9a2b-f14a6e38b568)

you can select FM by using switch SA
the GVAR values and so the channel outputs (and flap positions) are set directly without smooth transition time,
(here camber flap runs from 0(neutral) to 52% (max thermal)
so you"ll produce a kind of "hiccup" in models attitude 

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/20392683/c4ab1bf0-a978-4158-b404-7badfcb35edd)

there would a a workaround by using a "master GVAR" by using a range from -100 to 100 and using a slow function in the input, but then the user would have to do some maths 
(lets say user wants app. 50% deflection in a FM, he could set the master GV to 80% and the FM camber value to 62,5%
this way is too complicated for most users and could produce much failures

regards
udo

**Comment by BladeScraper-Designs** [comment]
I agree, Vars should fade just like other things do when switching flight modes.

**Comment by Jedsters** [comment]
I have similar problems with Fading not working as expected / desired.

Here's an interesting test.

Create a new model
Create 2 FM, SA Up and SA Middle 
Create a VAR called OFFVAR, add a value for FM0 of 15% and another value for FM1 of 30%
Create an OFFSET mixer Always on, FM0 with an offset source of OFFVAR output to channel 10
Clone that mixer and change the new mixer so it is for FM1

Now change between FM0 and FM1 and you'll see Ch10 immediately jumps between 15% and 30%. This isn't fading on channels as I'd expect.

This looks like when you flick the switch the FM fade calculation is done immediately before any VARs are updated so the logic says e.g. fade from 15% to 15% - job done and a few cycles later the VAR calculation is done and therefore the output channel jumps to it's new value of 30% since the 'fading' has already been completed. 

If this is what is happening then it would seem to be possible to adjust the code so that the fade calculation is done after VARs are updated. What the other implications for this may be I'm not sure.
 
Thanks

**Comment by Jedsters** [comment]
I should perhaps add that if 2 VARs are used with fixed values of 15% and 30% then that fades nicely as expected, indeed even if the VARs have values of FM0 15% for one VAR and FM1 30% for the other VAR then things still fade nicely even though the VAR values change according the the FM. If rather than using FM as the trigger you use the switch position, then things still work nicely. 
It's when you combine these two values into a single VAR that fading doesn't work properly. This seems a reasonable thing to want to do (indeed, a more likely thing).

Thanks

**Comment by RC-SOAR** [comment]
Agree, FM-based fades don't seem to be working as expected when VARs are injected into mixer weights, with the VARs themselves having FM conditional values.

**Comment by bsongis-frsky** [comment]
I don't think it's worth stopping the 1.5.0 process for this issue, I will look at it in 1.5.1

**Comment by Jedsters** [comment]
Agreed.

**Comment by glipski50** [comment]
Two examples from my side, just as additional information to support the topic.
In Ethos 1.4.x I used Multi-Offset-Mixer per flight mode activated be the defined flight modes. The fade is / fade out worked as defined.
![FM_Mischer](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/116186946/aadbdebf-ee78-4285-b99f-aaa3ba01af87)

Now in 1.5.1 I set only one Multi-Offset-Mixer for the flight modes and used VARS which had the values per flight mode defined. As already written above the fade in / fade out does not work that way.
![FM_Vars](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/116186946/fe7af0a1-e286-443a-8a95-e20cb9f5854c)

![Vars](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/116186946/a83d6ffe-b007-4675-861a-13d6d9f64f6e)

Without considering fade in / fade out that are defined in the flight mode menue the vars in 1.5.1 are not usable when programming flight modes. 

Here are also the two models files

[fm vars.zip](https://github.com/FrSkyRC/ETHOS-Feedback-Community/files/14521713/fm.vars.zip)

**Comment by Dsrenger** [comment]
I just tested this with my 1.4 models and non of them works after conversion to 1.5

In 1.4 I used VARs that had flight mode specific values, that got routed to channels. These channels were used in Offset mixers to apply flight mode specific offsets. Fade in/out values of the flightmodes were used to apply a smooth transition.
 
![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/51758752/4bdac28c-b248-42db-96b7-b1bf95007d52)

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/51758752/b641711b-919f-453d-bdcc-5658ccc74b6b)

When I open this model in 1.5.1 a VAR with the flight mode specific values is created:

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/51758752/60b67100-c8c7-456d-b3fa-4967806fb398)

As the 1.5 VAR cannot be routed to a channel ETHOS creates a free mix that routes the VAR to the channel

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/51758752/18e8bed1-14d3-45f3-ad79-6965aa19e8db)

Now the Offset Mixer gets its flight mode specific values from the channel
![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/51758752/020533b8-334e-4870-b2cd-49f95e9b42b9)

So far it is perfect! Good job!

The only thing, that is missing, is that the fade in/out from flight modes  get considered. The values dont have a smooth transition. The change hard from one value to the other.

Only solution is, to add a slow to the free mix that routes the VAR to the channel.

Currently I am not able to use the VAR directly in the offset mixer, as the Fade in/out from flightmodes are not considered there. At least the free mix is giving me the possibility to add a slow. 
But this feels not like the intended way to use VARs in 1.5.

Attched you will find a condensted model file (where the screenshot were taken from.

[model02.zip](https://github.com/FrSkyRC/ETHOS-Feedback-Community/files/14524221/model02.zip)

**Comment by Dsrenger** [comment]
Attached a model that shows the basic problem without any conversion specific stuff.

2 flight modes  with fade in times (1 second)
1 VAR with flight mode specific values
1 offset mixer that uses the VAR

When switching the flight mode the fade in times are not considered.

[test.zip](https://github.com/FrSkyRC/ETHOS-Feedback-Community/files/14524475/test.zip)

**Comment by strgaltdel** [comment]
today I remembered the general fade in/out discussions regarding FM switching.
I fear as more heli users move to 1.5.x, we will get the same results using vars without testing

maybe a hint is beneficial that var fading is already planned for 1.6
(question would be where to place?)

 https://github.com/FrSkyRC/ETHOS-Feedback-Community/issues/599
![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/20392683/c3b853a9-48bc-4cd4-b356-3a7c9b295c68)

the first issue / discussion regarding FM fading:
https://github.com/FrSkyRC/ETHOS-Feedback-Community/issues/93

**Comment by azaz44** [comment]
I just updated to 1.5 (. 2) and got this problem immediately. It's great that fade times can come from a source (I was requesting this change). But on the other hand fades don't work at all. I would suggest to treat it as critical. I can live with this because I don't plan to fly real models befor mid april or even may, but I think for many users this breaks existing functionality.

**Comment by azaz44** [comment]
I tell you what, I actually have to downgrade to 1.4.x because of this issue. There's quite a risk that I'll start flying soon in April, and for me this is quite a blocker. I don't want to make changes to models, which will prevent me to downgrade to an older version in future, and have to start a season without fading.

So I hope you can fix it much earlier that 1.6 which is added as a milestone here.

To make this comment useful I describe my case in quick points:
- I fly RC helis
- I use VARs (var mixers in 1.4.x) for things like throttle (rotor head RPM), aileron / elevator / rudder expo and rotation rates (weight) etc. (among others)
- I fly both 3D and F3C
- A difference in variable values between "F3C Hover" and "3D" can be super huge. For example, in "3D" mode, cyclic rotation rates are easily 5x faster than in "F3C hover" mode. Other parameters, like expo, have very big differences too. Hover mode is just so different.
- For this reason, rapid switch from "F3C Hover" to "3D" or the other way around would be very dangerous. You can loose control of the model. And this is where fading is used, to gradually bring you from one mode to the other in 2-4 seconds.
- I could of course not switch like this, but rather land first, switch the mode and take off again. But I have my habits and am used to switch it in flight, mix hovering and 3D in a single flight, finish the battery with hovering etc. And I don't want to crash a model because I forgot that I should not do it.

I hope this description adds something.

**Comment by glipski50** [comment]
Are there any new information, if, and if yes when, using vars in flight mides would include fade in and fade out. Above in this issue Bertrand included this into the 1.5.1 miles stones. But at the moment with 1.5.7 it is not included yet. 
@ Bertrand, please reconsider this topic in 1.6.0 where it is included at the moment.

**Comment by azaz44** [comment]
> today I remembered the general fade in/out discussions regarding FM switching. I fear as more heli users move to 1.5.x, we will get the same results using vars without testing
> 
> maybe a hint is beneficial that var fading is already planned for 1.6 (question would be where to place?)

Do you mean "where to place in UI"?
It is there in UI in flight modes, even in 1.5. It's just that it isn't working.

**Comment by azaz44** [comment]
Update: I try to use 1.5.x despite fading missing. I try to workaround the problem. The way to workaround is:
- map a var (VAR1 for example) to a "Free Mix" on some high channel number (like #40)
- use "Slow" action on the mixer
- use this #40 channel everywhere instead of a VAR
This is mostly easy to do, because Ethos will convert 1.4.x models exactly this way. For existing VAR mixers it will create a VAR but also keep a mixer of type FreeMix so in many cases it is just a matter of adding a SLOW action.

This will cause #40 to change value slowly after switching flight modes.

It is not perfect, because adjusting SLOW action parameters is different for every VAR and they will never behave perfectly. Mostly after switching FM one var will reach target value at a different time than the other. But for practical use it can work. And unfortunately it doesn't work for curves.

It certainly solves the problem of somebody just needs a solution for motor RPM.

Just in case it helps anybody. I just need changes in 1.5.x so much that I had to try something.

**Comment by azaz44** [comment]
@bsongis-frsky About curves, this is something to consider when fixing this problem.

I have such setup for Pitch channel:

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/42830563/9a5c2048-838d-40e5-8b8d-40f12b5fd133)

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/42830563/5137fe11-bdf3-4d2b-b0bd-466f5893f458)

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/42830563/0442c494-6283-4790-8dba-c4c314b920c3)

In short:
- there is a "Curve" action
- in "Hover" FM, it is a custom curve ("Hover Pitch")
- in all other modes it is just some expo

It's a direct conversion from what was in 1.4.x.
And this worked in 1.4.x, I mean it did use fading.

So the problem is not only with fading VARs but fading curves too. And maybe there are other places too.

As of now, this also doesn't work in 1.5.7. If I switch to FM4-Hover, I get a sudden jump of pitch, because it switches to a different curve immediately without fading.

Just wanted to throw this in so it's not forgotten. I just realized the whole FM fading is more complex than I thought.

**Comment by azaz44** [comment]
Unstale please, very important and awaited feature

**Comment by Dsrenger** [comment]
My personal highest prio item. 
Will remove the stale label until the end of the days ;)

**Comment by azaz44** [comment]
This is super needed, using ugly workarounds all the time.

**Comment by Dsrenger** [comment]
As we are heading towards 1.7 I try my luck again

For me, this is my last issue with Ethos in my "daily" programming work.

Using VARs for an offset mixer to set flaps FM specific is so elegant. But then having to route the VAR through a channel to apply the slow is extremly unintuitive.

VARs need to be considered with FM fade in/out

I recently had somebody switching from Jeti to FrSky and he had a very hard time to understand, that the flaps are moving instantly considering he had a FM fade in/out defined.

Please think about to add this to 1.7

Thanks
Dennis

**Comment by lthole** [comment]
@azaz44 Can you please post your model incorporating the work-arounds you needed to overcome the fade out/in issues? I am trying to get a better understanding of the issue.

**Comment by Jedsters** [comment]
I was having a look at @azaz44 suggestion of using extra channels and it does seem to be a bypass. 

Create a basic 4 channel model.
Add a FM of Flaps active with SA Down and give it and the default FM fade in / out values of e.g. 1s.
Add a VAR of Flap Weight with default value -20% and FM1 (Flaps) value of 30%
Add a Camber mix, active in Flight Mode Flaps only, Assign VAR Flap Weight to Output1 Weight. Change Output2 Weight to 30% to make it easier to compare with a non VAR.

Now create a Free mix active in all FM with a source of _Flap Weight_ and add a action of Always on Slow with fade down of 1s and output to e.g. Ch20. 
Clone the above Camber mix and edit it. Use Ch20 for the Output1 Weight and output to ch7, also adjust Output2 to Ch10. 

Now go to Outputs
Channels 1 demonstrate the original problem...
With SA middle channel 1 and channel 5 are 0%
Move SA Down which causes ch1 and ch5 to fade up to -30%
Now move SA middle which causes ch1 to instantly jump to 20% and fade down from there instead of fading up from -30%. By contrast ch5 correctly fades up from -30%

Channel 7 demonstrates the bypass
With SA middle channel 7 is 0%
Move SA Down which causes ch7 to fade down to -30%
However move SA middle causes ch7 to quickly fade up to maybe 12% and then fade down from there to 0%. So long as the Var you are switching to has greater magnitude and is negative of the value being switched from this seems to be the case.
However, if you now change the Mixer fade down value to 10s then things improve and look ok. 

I suspect what is happening is the Mixer is smoothing out the sudden transition in my example in the opposite direction, but I'm not sure it will work in all circumstances (I'm not saying it won't, just I don't know). I'm guessing there is some calculation that can be done involving the percentages, fade time and slow time to work out what slow time is neede, or maybe 25s can always be used?

While it may be a clever bypass I'd still like to understand it this problem will be fixed or not. If it will be fixed then this bypass could be a useful interim solution, however if it isn't going to be fixed then I'll probably adopt another solution (increased number of mixes). This was obviously first raised 2.5yrs ago and scheduled for 26.x with no due date. 

As for curves, I've not got my head around that yet.

**Comment by lthole** [comment]
I tried your example. You seem to start with the Var at -20 and +30, but further down you seem to have reversed that. In my example ch7 starts at +30 and jumps to -12 and then fades to 0.
I think this is the wrong approach. Until this is fixed leave FM fade in/out at 0 and control all fading from the high mixes.

I am going to investigate fading between helicopter throttle curves.

**Comment by Jedsters** [comment]
Oh, apologies, I've mixed things up while typing...

**Comment by lthole** [comment]
I modified the example:
1 Set FM  half flaps SA mid, full flaps SA down, fades to 0
2. Set Var to default-20%, FM1 half flaps 30% and FM2 full flaps 60%.
3. Create a high mix with all FM’s on, slow up 2s, slow down 2s, weight = var, output to ch20.
4. Create a camber mix with all FM’s on, source = MAX,  output1 weight = ch20.
Works butter smooth up and down.

**Comment by Jedsters** [comment]
If I understand what you've done, you've possibly broken other things?
The FM fades aren't necessarily only fading one or a few mixes, they could be fading other things which work correctly because they don't use VARs as Output weights. 
If you remove FM fades then all of these other things will also now need to go via extra channels which is far from ideal. One of the things I liked about ethos compared to OTX was that I didn't have any interim channels, this is adding a load back in which just complicates things.  
With the previous method the FM fades continued to work elsewhere and the mixes which it didn't had a workaround. 
Or have I misunderstood when you said  Set FM half flaps SA mid, full flaps SA down, fades to 0 you didn't mean change them all to 0secs?

**Comment by lthole** [comment]
Yes if you have flight mode dependent Vars or curves that need slow, I think you have to set FM fades to 0, and do all slows in high mixes. How many things really need slow? I know this is suboptimal, but I believe this will be fixed. In the meantime at least you won’t run out of resources doing the work-arounds.

**Comment by Dsrenger** [comment]
You really have to set fades to 0

It really gets extremely fizzy when you have flight mode specific trims.

Because the Vars jumps immediately and the trims will be applied slowly....a use case you forget during testing your model

I had a near crash when I applied FM specific offsets on an elevator and used FM trim on the elevator.

The elevator jumped to the new position and then slowly back to the trimmed position.

**Comment by Jedsters** [comment]
We seem to be 'fixing' the originally described problem by saying we should stop using FM fade if it causes problems. If this isn't going to be fixed I'd (almost) rather see VARs removed from mix outputs - but that would stop the above bypass which I'm sure some would find useful. Maybe the manual needs to be updated to address the problem explaining the problem in the absence of a fix. 

The problem seems to be a momentary glitch due to a VAR changing value resulting in mix outputs suddenly changing value as a result of changes before the fading is applied.

I'm not sure it would fix all scenarios, but my problem seems to stem from the change of VAR resulting in a sudden change to Output when the Mix is in any case becoming inactive!  This is just wrong, if the mix is becoming inactive then why are any changes to outputs as a result of VAR changes even being calculated? It should just fade from it's last active value. For starters maybe this could be corrected so that a mix becomes inactive before any new Output from the VAR change is calculated? 

Beyond this I'm struggling to work out how things can work other than the above bypass. Would another solution be to internalise the extra channel processing so we don't have to do anything?

My interim solution is to not use VARs on some mixes, but that means I need a load more mixes to do that as every VAR in an mix needs to be converted to a fixed values. But at least FM fade everything correctly. I've also had to simplify my setup a bit to reduce the number of mixes needed, but I wasn't really using a couple much anyway so that's no big loss.

**Comment by mawzthefinn** [comment]
I'd really like this to get fixed, it's a poor interaction of two of ETHOS's most powerful features and as a result I've limited my use of FM's to preserve VAR functionality.
